src/main/java/net/openhft/hashing/MurmurHash_3.java

```diff
@@ -205,8 +205,15 @@ private static long finalize(long length, long h1, long h2) {
         h1 = fmix64(h1);
         h2 = fmix64(h2);
 
-        h1 += h2;
-        return h1;
+
+        if (null != result) {
```

Comment:
some c hash api use this style. with a nullable container, we can easily implement low 64-bit hash or full hash in one function with very little runtime cost. When Tuple Hash extends Long Hash, this  is a compromise for easy development and least allocation.

if a pure tuple hash function, then we only need the api like ones in the apache-commons:

```java
long[] hash(..., long[] non_null_array); // pre alloc
long[] hash(...); // alloc on the fly
```
they are all need an additional allocation to become a 64-bit hash.

@leventov which do you prefer?

Reply:
The internal implementation may be the same as it is right now (i. e. shared 128/64 bits implementation where it is possible), but on the public API level, I think the best approach is:
 1. Having `LongTupleHashFunction` as a separate class, not extending `LongHashFunction`
 2. This class having methods `void hash(..., long[] non_null_output_array)`. `void` return emphasizes that the last argument is the output parameter, preventing any possible confusion, such as somebody accidentally calling `long result = f.hashInt(len, array);` and assuming that the contents of the `array` should be hashed till `len`, whereas it will actually just hash the 32-bit value `len` into the output array.

Reply:
> `long result = f.hashInt(len, array)`

This is a good case~

Given LongTupleHashFunction separated from LongHashFunction, the shared implementation will be a soft suggestion for the developers. When creating a function instance, the user should explicitly select LongHashFunction or LongTupleHashFunction. The javadoc of 128 bits hash should also tell which part of the results is returned from a LongHashFunction instance. 

The new api and an example implementation would be:

```java
class LongTupleHashFunction {
  public int resultBits();
  public long[] newResultArray();

  public void hashXXX(input, @NonNull long[] result); // reuse pre-allocation
  public long[] hashXXX(input); // allocation on-the-fly, is this needed?
}

class Murmur extends LongTupleHashFunction {
  // extending LongTupleHashFunction is just a sugar for `asTupleHash()`

  // for internal only,
  // can combine the 64/128 bits implementation by check whether `result == null`
  protected abstract long hashXXX(input, long[] result);

  public static LongTupleHashFunction asTupleHash();
  public static LongHashFunction asLongHash();

}
```

Reply:
New API that you suggested looks good to me.

Reply:
do u think the alloc-on-the-fly api is needed? it may always return a same array.

Reply:
You could add it. Since the garbage-free methods are `void`, I don't see how these overloads can lead to any bugs or confusion for users.

However, the methods must not return the same array, because array contents are mutable and users may modify them after they obtained the arrays. Memoizing and returning the same array would be a surprising and bug-prone here. These methods should just allocate.